src/infrastructure/server/routers/APIRouter.ts

```diff
@@ -104,14 +95,32 @@ export default class APIRouter {
 }
 
 const handler =
-  (action: ControllerAction) => async (req: Request, res: Response) => {
+  (action: ControllerAction, route: RouteDefinition) =>
+  async (req: Request, res: Response) => {
     try {
-      const params = { ...req.query, ...req.body };
-      const response = await action(params /*, req.User */);
+      console.log(req.cookies)
+      const params = (route.parameters || []).map(({ in: from, ParameterType: Type }) => {
+        switch (from) {
+          case "query":
+            return new Type(queryString.parseUrl(req.url, { parseBooleans: true, parseNumbers: true }) || {});
+          case "cookie":
+            return new Type(req.cookies || {});
+          case "header":
+            return new Type(req.headers || {});
+          // return Object.keys(req.headers || {}).reduce((headers, key) => ({...headers, [key]: req.headers[key]}), {});
+          case "path":
+            return new Type(req.params || {});
+          case "body":
+          default:
+            return new Type(req.body || {});
+        }
+      });
+      const response = await action(...params /*, req.User */);
 
       return res.status(response.status).json(response);
     } catch (e) {
-      const response = HttpResponse.internalError(e.message);
+      Logger.error(e);
+      const response = HttpResponse.internalError(e.message || e.toString());
 
       return res.status(response.status).json(response);
     }
```

Comment:
Find a better place for the handler wrapper. 

Should I implement polymorphism?